internal/encoding/exported_types.go

```diff
@@ -0,0 +1,425 @@
+package encoding
+
+import (
+	"encoding/hex"
+	"errors"
+	"fmt"
+	"math"
+
+	"github.com/Azure/go-amqp/internal/buffer"
+)
+
+// Durability Policies
+const (
+	// No terminus state is retained durably.
+	DurabilityNone Durability = 0
+
+	// Only the existence and configuration of the terminus is
+	// retained durably.
+	DurabilityConfiguration Durability = 1
+
+	// In addition to the existence and configuration of the
+	// terminus, the unsettled state for durable messages is
+	// retained durably.
+	DurabilityUnsettledState Durability = 2
+)
+
+// Durability specifies the durability of a link.
+type Durability uint32
+
+func (d *Durability) String() string {
+	if d == nil {
+		return "<nil>"
+	}
+
+	switch *d {
+	case DurabilityNone:
+		return "none"
+	case DurabilityConfiguration:
+		return "configuration"
+	case DurabilityUnsettledState:
+		return "unsettled-state"
+	default:
+		return fmt.Sprintf("unknown durability %d", *d)
+	}
+}
+
+func (d Durability) Marshal(wr *buffer.Buffer) error {
+	return Marshal(wr, uint32(d))
+}
+
+func (d *Durability) Unmarshal(r *buffer.Buffer) error {
+	return Unmarshal(r, (*uint32)(d))
+}
+
+// Expiry Policies
+const (
+	// The expiry timer starts when terminus is detached.
+	ExpiryLinkDetach ExpiryPolicy = "link-detach"
+
+	// The expiry timer starts when the most recently
+	// associated session is ended.
+	ExpirySessionEnd ExpiryPolicy = "session-end"
+
+	// The expiry timer starts when most recently associated
+	// connection is closed.
+	ExpiryConnectionClose ExpiryPolicy = "connection-close"
+
+	// The terminus never expires.
+	ExpiryNever ExpiryPolicy = "never"
+)
+
+// ExpiryPolicy specifies when the expiry timer of a terminus
+// starts counting down from the timeout value.
+//
+// If the link is subsequently re-attached before the terminus is expired,
+// then the count down is aborted. If the conditions for the
+// terminus-expiry-policy are subsequently re-met, the expiry timer restarts
+// from its originally configured timeout value.
+type ExpiryPolicy Symbol
+
+func ValidateExpiryPolicy(e ExpiryPolicy) error {
+	switch e {
+	case ExpiryLinkDetach,
+		ExpirySessionEnd,
+		ExpiryConnectionClose,
+		ExpiryNever:
+		return nil
+	default:
+		return fmt.Errorf("unknown expiry-policy %q", e)
+	}
+}
+
+func (e ExpiryPolicy) Marshal(wr *buffer.Buffer) error {
```

Comment:
If I had to do this over again, I'd refactor the types such that the [un]marshal methods aren't methods but internal helper funcs. Then they're not part of public surface area. However, since we've already leaked some of them, and `buffer.Buffer` isn't publicly exported, I don't think it's worth it to fix some and not all especially since fixing the already exported ones is a breaking change.

Reply:
Agreed.